rohmu/object_storage/s3.py

```diff
@@ -592,6 +596,93 @@ def check_or_create_bucket(self) -> None:
+    ) -> ConcurrentUploadId:
+        args: dict[str, Any] = {
+            "Bucket": self.bucket_name,
+            "Key": self.format_key_for_backend(key, remove_slash_prefix=True),
+        }
+        if metadata:
+            metadata = self.sanitize_metadata(metadata)
+            args["Metadata"] = metadata
+        if self.encrypted:
+            args["ServerSideEncryption"] = "AES256"
+        if mimetype:
+            args["ContentType"] = mimetype
+
+        self.stats.operation(StorageOperation.create_multipart_upload)
+        try:
+            cmu_response = self.s3_client.create_multipart_upload(**args)
+        except botocore.exceptions.ClientError as ex:
+            raise StorageError("Failed to initiate multipart upload for {}".format(key)) from ex
+
+        upload = ConcurrentUploadData("aws", cmu_response["UploadId"], key)
+        upload_id = self._new_concurrent_upload(upload, metadata)
+        return upload_id
+
+    def complete_concurrent_upload(self, upload_id: ConcurrentUploadId) -> None:
+        concurrent_data, metadata, chunks = self._get_concurrent_upload(upload_id)
+        backend_key = self.format_key_for_backend(concurrent_data.key, remove_slash_prefix=True)
+        sorted_chunks = sorted(
+            ({"ETag": etag, "PartNumber": number} for number, etag in chunks.items()),
+            key=lambda part: cast(int, part["PartNumber"]),
+        )
+        try:
+            self.s3_client.complete_multipart_upload(
+                Bucket=self.bucket_name,
+                Key=backend_key,
+                MultipartUpload={"Parts": sorted_chunks},  # type: ignore[typeddict-item]
+                UploadId=concurrent_data.backend_id,
+                RequestPayer="requester",
+            )
+            del self._concurrent_uploads[upload_id]
+            # NOTE: metadata is not available if we are resuming. size is never available.
+            self.notifier.object_created(key=backend_key, size=None, metadata=metadata)
+        except botocore.exceptions.ClientError as ex:
+            raise StorageError("Failed to complete multipart upload for {}".format(concurrent_data.key)) from ex
+
+    def abort_concurrent_upload(self, upload_id: ConcurrentUploadId) -> None:
+        concurrent_data, _, _ = self._get_concurrent_upload(upload_id)
+        backend_key = self.format_key_for_backend(concurrent_data.key, remove_slash_prefix=True)
+        try:
+            self.s3_client.abort_multipart_upload(
+                Bucket=self.bucket_name,
+                Key=backend_key,
+                UploadId=concurrent_data.backend_id,
+                RequestPayer="requester",
+            )
+            del self._concurrent_uploads[upload_id]
+        except botocore.exceptions.ClientError as ex:
+            raise StorageError("Failed to abort multipart upload for {}".format(concurrent_data.key)) from ex
+
+    def upload_concurrent_chunk(
```

Comment:
I don't think we can reuse much. The other method has a key difference: it decides the chunk size.

Here we are given a `BinaryIO` by the user which needs to be uploaded as a single call to S3. We cannot read the whole file into memory because it could be up to 5GiB and the user might call these method concurrently in multiple threads so RAM would explode. That's why I created that `ProgressStream` that wraps the original one and simply records how much bytes were read.

Retries are missing yes.

Reply:
`multipart_upload_file_object` could split at the point where `data` becomes available. 

it uses self._read_bytes which reads `self.multipart_chunk_size` bytes of data into memory